test/jdk/java/net/httpclient/ContentLengthHeaderTest.java

```diff
@@ -0,0 +1,226 @@
+                .uri(URI.create(testContentLengthURI + BODY_PATH))
+                .build();
+        HttpClient hc = HttpClient.newHttpClient();
+        HttpResponse<String> resp = hc.send(req, HttpResponse.BodyHandlers.ofString(UTF_8));
+        assertEquals(resp.statusCode(), 200, resp.body());
+    }
+
+    @Test
+    // A HEAD request with no request body should have no Content-length header
+    public void headWithNoBody() throws IOException, InterruptedException {
+        testLog.println("Checking HEAD with no request body");
+        HttpRequest req = HttpRequest.newBuilder()
+                .version(HTTP_1_1)
+                .HEAD()
+                .uri(URI.create(testContentLengthURI + NO_BODY_PATH))
+                .build();
+        HttpClient hc = HttpClient.newHttpClient();
+        HttpResponse<String> resp = hc.send(req, HttpResponse.BodyHandlers.ofString(UTF_8));
+        assertEquals(resp.statusCode(), 200, resp.body());
+    }
+
+    @Test
+    // A HEAD request with a request body should have a Content-length header
+    public void headWithBody() throws IOException, InterruptedException {
+        testLog.println("Checking HEAD with request body");
+        HttpRequest req = HttpRequest.newBuilder()
+                .version(HTTP_1_1)
+                .method("HEAD", HttpRequest.BodyPublishers.ofString("HEAD Body"))
+                .uri(URI.create(testContentLengthURI + BODY_PATH))
+                .build();
+        HttpClient hc = HttpClient.newHttpClient();
+        // Sending this request invokes sendResponseHeaders which emits a warning about including
+        // a Content-length header with a HEAD request
+        HttpResponse<String> resp = hc.send(req, HttpResponse.BodyHandlers.ofString(UTF_8));
+        assertEquals(resp.statusCode(), 200, resp.body());
+    }
+
+    public static void handleResponse(HttpExchange ex, String body, int rCode) throws IOException {
+        try (OutputStream os = ex.getResponseBody()) {
+            byte[] bytes = body.getBytes(UTF_8);
+            ex.sendResponseHeaders(rCode, bytes.length);
+            os.write(bytes);
+        }
+    }
+
+    static class NoContentLengthHandler implements HttpHandler {
+
+        @Override
+        public void handle(HttpExchange exchange) throws IOException {
+            testLog.println("NoContentLengthHandler: Received Headers " + exchange.getRequestHeaders().entrySet() +
+                            " from " + exchange.getRequestMethod() + " request.");
+            String contentLength = exchange.getRequestHeaders().getFirst("Content-length");
+
+            // Check Content-length header was not set
+            if (contentLength == null) {
+                handleResponse(exchange, "Request completed",200);
+            } else {
+                String responseBody = exchange.getRequestMethod() + " request contained an unexpected " +
+                        "Content-length header.";
```

Comment:
Maybe the message could include the value of `Content-Length` that was received.

Reply:
Also it's always better to drain the request input stream even if there is no bytes.

Reply:
You could possibly do that in `handleResponse()`